core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ArrayBindingBasedQueryEvaluationContext.java

```diff
@@ -61,23 +94,58 @@ public ArrayBindingSet createBindingSet() {
 
 	@Override
 	public Predicate<BindingSet> hasBinding(String variableName) {
-		Function<ArrayBindingSet, Boolean> directHasVariable = new ArrayBindingSet(allVariables)
-				.getDirectHasBinding(variableName);
-		return (bs) -> {
+		if (initialized) {
+			for (int i = 0; i < allVariables.length; i++) {
+				if (allVariables[i] == variableName) {
+					return hasBinding[i];
+				}
+			}
+		}
+
+		assert variableName != null && !variableName.isEmpty();
+		Function<ArrayBindingSet, Boolean> directHasVariable = defaultArrayBindingSet.getDirectHasBinding(variableName);
+		return new HasBinding(variableName, directHasVariable);
```

Comment:
Using a class here instead of a lambda helps the compiler for some reason. I think the issue might be because we had one lambda calling another lambda.

Reply:
Interesting, I had not expected that. This makes me currios which JDK version are you testing on?

Reply:
18 on MacOS aarch64